cmd/soroban-cli/src/rpc/transaction.rs

```diff
@@ -84,6 +92,189 @@ pub fn assemble(
+    // already in the auth?
+    // let ledger_validity_count = 5;
+
+    let source_address = source_key.public.as_bytes();
+
+    let signed_auths = body
+        .auth
+        .iter()
+        .map(|raw_auth| {
+            let mut auth = raw_auth.clone();
+            let SorobanAuthorizationEntry {
+                credentials: SorobanCredentials::Address(ref mut credentials),
+                ..
+            } = auth else {
+                // Doesn't need special signing
+                return Ok(auth);
+            };
+            let SorobanAddressCredentials { ref address, .. } = credentials;
+
+            // See if we have a signer for this authorizationEntry
+            // If not, then we Error
+            let needle = match address {
+                ScAddress::Account(AccountId(PublicKey::PublicKeyTypeEd25519(Uint256(ref a)))) => a,
+                ScAddress::Contract(Hash(c)) => {
+                    // This address is for a contract. There's no way to sign it because it is malformed.
+                    return Err(Error::MissingSignerForAddress {
+                        address: stellar_strkey::Strkey::Contract(stellar_strkey::Contract(*c))
+                            .to_string(),
+                    });
+                }
+            };
+            let signer = if let Some(s) = signers.iter().find(|s| needle == s.public.as_bytes()) {
+                s
+            } else if needle == source_address {
+                // This is the source address, so we can sign it
+                source_key
+            } else {
+                // We don't have a signer for this address
+                return Err(Error::MissingSignerForAddress {
+                    address: stellar_strkey::Strkey::PublicKeyEd25519(
+                        stellar_strkey::ed25519::PublicKey(*needle),
+                    )
+                    .to_string(),
+                });
+            };
+
+            sign_soroban_authorization_entry(
+                raw_auth,
+                signer,
+                signature_expiration_ledger,
+                &network_id,
+            )
+        })
+        .collect::<Result<Vec<_>, Error>>()?;
+
+    body.auth = signed_auths.clone().try_into()?;
+    tx.operations = vec![op].try_into()?;
+    Ok((tx, signed_auths))
+}
+
+pub fn sign_soroban_authorization_entry(
+    raw: &SorobanAuthorizationEntry,
+    signer: &ed25519_dalek::Keypair,
+    signature_expiration_ledger: u32,
+    network_id: &Hash,
+) -> Result<SorobanAuthorizationEntry, Error> {
+    let mut auth = raw.clone();
+    let SorobanAuthorizationEntry {
+        credentials: SorobanCredentials::Address(ref mut credentials),
+        ..
+    } = auth else {
+        // Doesn't need special signing
+        return Ok(auth);
+    };
+    let SorobanAddressCredentials { nonce, .. } = credentials;
+
+    let preimage = HashIdPreimageSorobanAuthorization {
```

Comment:
This should be wrapped into `HashIdPreimage::SorobanAuthorization` (thisi is to ensure we have a seprate 'namespace' for all the signed preimages core might deal with.